File: Algorithms/src/110. Balanced Binary Tree/Code/solution.cpp
class Solution
{
public:
    // Helper function that returns height of the tree
    // If tree is unbalanced, return -1
    int height(TreeNode *root)
    {
        if (root == nullptr)
            return 0;

        int leftHeight = height(root->left);
        if (leftHeight == -1)
            return -1;

        int rightHeight = height(root->right);
        if (rightHeight == -1)
            return -1;

        if (abs(leftHeight - rightHeight) > 1)
            return -1;

        return 1 + max(leftHeight, rightHeight);
    }

    bool isBalanced(TreeNode *root)
    {
        return height(root) != -1;
    }
};

File: Algorithms/src/110. Balanced Binary Tree/Code/solution.go
func isBalanced(root *TreeNode) bool {

    var height func(node *TreeNode) int
    height = func(node *TreeNode) int {
        if node == nil {
            return 0
        }

        leftHeight := height(node.Left)
        if leftHeight == -1 {
            return -1
        }

        rightHeight := height(node.Right)
        if rightHeight == -1 {
            return -1
        }

        if leftHeight-rightHeight > 1 || rightHeight-leftHeight > 1 {
            return -1
        }

        if leftHeight > rightHeight {
            return 1 + leftHeight
        }
        return 1 + rightHeight
    }

    return height(root) != -1
}
